sd.cpp: add samplers

vmobilis opened the pull request:
The samplers Euler and Euler Ancestral are separated in their own branches of samplers switch.

For Euler, I was referring to `stable-diffusion.cpp`: https://github.com/leejet/stable-diffusion.cpp/blob/10c6501bd05a697e014f1bee3a84e5664290c489/denoiser.hpp#L535

Euler Ancestral is left unchanged.

Example (for same model, prompt and steps):
<table>
<tr><th width="50%">Euler Ancestral</th><th width="50%">Euler</th></tr>
<tr><td><img src="https://github.com/user-attachments/assets/6cadd5fe-70c4-45d4-af5e-bdc913db93f3"></td>
<td><img src="https://github.com/user-attachments/assets/c61ad207-3205-4fb9-b816-73c7280f6a2a"></td></tr>
</table>

And thank you for the program.

vmobilis committed:
sd.cpp: moving steps number to separate variable

It will be needed for Euler sampler.

vmobilis committed:
sd.cpp: adding --sampler option

Also, moving time counter after a sampler.

vmobilis committed:
sd.cpp: rearranging lines to view next commit more easily

vmobilis committed:
sd.cpp: creating samplers switch

Also, using x_ptr as counter and removing hw variable, to simplify

vmobilis committed:
sd.cpp: adding Euler sampler

vmobilis committed:
sd.cpp: tried to simplify sampling algorithms

Alternative formulas noticeably change image and therefore are turned off.

vmobilis commented:
@vitoplantamura, hello, I slightly changed the sampling cycle: instead of using hw counter, I'm simply checking x_ptr, this approach removes 1 increment operation.
The changes are visible here: https://github.com/vitoplantamura/OnnxStream/commit/523bc53cab6d780a5d2b03c72b9d378f6f37c714#diff-5cb305592bc405473a9934a751e61fa93aafa9d3b2f6bd67de62af51ab53e4d0R1346.

Also, I tried to simplify sampling formula a little, to speed up, but speed change is insignificant, so maybe last commit (https://github.com/vitoplantamura/OnnxStream/pull/120/commits/01bfcfcf2c200cfa9da119d902472d7130f9d616) is not necessary.
It's turned off, but you can check the differences by setting [ORIGINAL_SAMPLER_ALGORITHMS](https://github.com/vitoplantamura/OnnxStream/pull/120/files#diff-5cb305592bc405473a9934a751e61fa93aafa9d3b2f6bd67de62af51ab53e4d0R1333) to 0.

vmobilis committed:
README.md: description of --sampler option

vmobilis commented:
@vitoplantamura, I tested some more, it looks like it does not work for SDXL Turbo model.

vitoplantamura commented:
very cool!

My guess is that if it works with SD1.5 and SDXL it should work with SDXL Turbo...

Vito

vmobilis committed:
Merge branch 'vitoplantamura:master' into patch-2

vmobilis committed:
sd.cpp: add DPM++ (2M) sampler

vmobilis commented:
@vitoplantamura, tried dpm++2m (updated examples above), something is definitely wrong.

vitoplantamura commented:
mmmh, maybe it's something to do with the fact that SDXL Turbo doesn't support negative prompts...

Vito

vmobilis committed:
sd.cpp: add DPM++ (2S) Ancestral sampler

vmobilis commented:
@vitoplantamura, DPM++ (2S) Ancestral [works better](https://github.com/vitoplantamura/OnnxStream/pull/120#issuecomment-2833484459), but Turbo model still does not look like a photo.

vmobilis committed:
sd.cpp: correct sigmas for Turbo model

vmobilis commented:
@vitoplantamura, I think I found a workaround:
https://github.com/vitoplantamura/OnnxStream/blob/d5374bf029f3bd0baf7036e3353eb9d87e1455a8/src/sd.cpp#L1420-L1430
The problem looks related to noise scheduler, by slightly lowering sigma values, it's possible to get more acceptably looking results:
<table><tr>
<td>
<img src="https://github.com/user-attachments/assets/9fe4980c-b526-4304-867b-4f54f8d6ccb2">
</td>
<td>
<img src="https://github.com/user-attachments/assets/d6d16cf4-00a5-4424-8f75-f4ab6b34a6f0">
</td>
</tr></table>

I also applied Euler sampler at last step, as recommended here: https://github.com/huggingface/diffusers/pull/5541, but applying only it does not fix images.

So for all non-ancestral samplers I'll try to use corrected sigma `si1` instead of `sigma[i + 1]`. So far it worked for all samplers I tried.

vmobilis committed:
sd.cpp: add DPM2 sampler

vmobilis committed:
sd.cpp: add LCM sampler

vmobilis committed:
sd.cpp: separate modified and normal Euler, to clarify

vmobilis committed:
sd.cpp: add Heun sampler

vmobilis committed:
sd.cpp: speedup DPM++2M

By using `a += b * c` instead of `a -= b * c`, compiler can utilize fused multiply + add when possible.

vmobilis committed:
sd.cpp: sampler names are unique

vmobilis committed:
Merge branch 'vitoplantamura:master' into patch-2

vmobilis committed:
sd.cpp: add DPM++ 2M v2 sampler

vmobilis committed:
sd.cpp: add iPNDM samplers

vmobilis committed:
sd.cpp: add variable step iPNDM

SD.cpp seems to use simplified form of iPNDM_v, added original version.

vmobilis committed:
sd.cpp: add taylor3 sampler

vmobilis committed:
sd.cpp: add dpm++2s sampler

Non-ancestral version of dpm++2s_a.

vmobilis committed:
sd.cpp: taylor3 also needs lighter scheduler correction

Same as for dpm++2mv2.

vmobilis committed:
sd.cpp: add DDPM samplers

vmobilis committed:
sd.cpp: add DDIM sampler

vmobilis committed:
sd.cpp: add TCD samplers

vmobilis committed:
sd.cpp: add DDIM Ancestral sampler

vmobilis committed:
sd.cpp: add DPM++ (3M) SDE sampler

vmobilis committed:
sd.cpp: add linear multi-step sampler

vitoplantamura commented:
hi @vmobilis,

I was playing with your latest commit and found that it didn't compile (on Fedora 40, GCC 14.2.1). The problem is the auto keyword not being able to deduce the same type (between float and double) in the comma-separated statements, and to a missing std::max overload. This is my patch:

```patch
diff --git a/src/sd.cpp b/src/sd.cpp
index 87ba151..eb145f1 100644
--- a/src/sd.cpp
+++ b/src/sd.cpp
@@ -2131,7 +2131,7 @@ inline static ncnn::Mat diffusion_solver(int seed, int step, const ncnn::Mat& c,
                     std::srand(seed++);
                     ncnn::Mat randn = randn_4_w_h(rand() % 1000, g_main_args.m_latw, g_main_args.m_lath);
                     const auto autozero = si1 * 0,
-                        variance = si1 * sqrt(std::max(autozero, 1 - pow(si1 / sigma[i], 2 * eta)));
+                        variance = si1 * std::sqrt(std::max(autozero, 1 - std::pow(si1 / sigma[i], 2 * eta)));
                     for (int c = 0; c < 4; c++)
                     {
                         float* x_ptr = x_mat.channel(c); const float* x_ptr_end = x_ptr + latent_length;
@@ -2799,11 +2799,11 @@ inline static ncnn::Mat diffusion_solver(int seed, int step, const ncnn::Mat& c,
                     for (; x_ptr < x_ptr_end; x_ptr++)
                     {
                         const auto model_output = (autozero + *x_ptr - *d_ptr++) / sigma[i],
-                        pred_original_sample = (*x_ptr * sqrt(alpha_prod_t)
-                            - model_output * sqrt(beta_prod_t)) / sqrt(alpha_prod_t),
+                        pred_original_sample = (*x_ptr * std::sqrt(alpha_prod_t)
+                            - model_output * std::sqrt(beta_prod_t)) / std::sqrt(alpha_prod_t),
                         pred_sample_direction = model_output
-                            * sqrt(1 - alpha_prod_t_prev - variance * eta * eta);
-                        *x_ptr = sqrt(alpha_prod_t_prev) * pred_original_sample
+                            * std::sqrt(1 - alpha_prod_t_prev - variance * eta * eta);
+                        *x_ptr = std::sqrt(alpha_prod_t_prev) * pred_original_sample
                             + pred_sample_direction;
                     }
                 }
@@ -2892,10 +2892,10 @@ inline static ncnn::Mat diffusion_solver(int seed, int step, const ncnn::Mat& c,
                 if (eta > 0 && i < (steps - 1)) {
                     std::srand(seed++);
                     ncnn::Mat randn = randn_4_w_h(rand() % 1000, g_main_args.m_latw, g_main_args.m_lath);
-                    const auto a = sqrt(alpha_n / alpha_s),
+                    const auto a = std::sqrt(alpha_n / alpha_s),
                     // alpha_s can be smaller than alpha_n because of sigma_reshaper() or different noise scheduler,
                     // therefore trimming negative differences
-                    autozero = si * 0, b = sqrt(std::max(autozero, 1 - alpha_n / alpha_s));
+                    autozero = si * 0, b = std::sqrt(std::max(autozero, 1 - alpha_n / alpha_s));
                     for (int c = 0; c < 4; c++)
                     {
                         float* x_ptr = x_mat.channel(c);
```

Vito

vmobilis committed:
sd.cpp: using std namespace for min, max, log, exp, sqrt, pow, abs

vmobilis commented:
@vitoplantamura, thank you, I suspected that assumption "who cares about difference in sqrt() and std::sqrt()" could be wrong... 😄
I think it should compile now.

vmobilis committed:
sd.cpp: add Gaussian integrals to LMS sampler

vitoplantamura commented:
Yes, with your fix, it compiles now 😀

I just committed a fix for the --ram option: it works now. If you have
enough RAM, specifying --ram should give you a performance boost (for the
UNET model) of about 20% (rough estimate and highly variable depending on
the architecture, SSD speed, etc.). It should be close to the maximum speed
for the architecture.

Vito

vmobilis commented:
@vitoplantamura, I tried with destructor and ram fixes, 1st works, with second application crashes.

It works:
- without `--ram` option;
- with `--ram`, but when I use `--ops-printf`.

Checked with SD 1.5 and Dreamshaper 8 models.

vmobilis committed:
Merge branch 'vitoplantamura:master' into patch-2

vmobilis committed:
sd.cpp: using 15k divisions for integrals, 120k is too slow

vitoplantamura commented:
Thanks for testing it. I've been doing some more tests with the `--ram` option in the meantime. The reason the speedup is only 15% is due to the involvement of the OS's disk cache when `--ram` is not used (and when there is enough free memory). Interestingly, I found that `--ram` can be slower if there is little memory available: my guess is that, in this case, OnnxStream's `DiskPrefetchWeightsProvider`'s disk reading pattern (which is sequential and ahead of time) is more performant than the OS's reading of cached-out memory pages from disk (with `--ram` and no physical memory available)...

Vito

vmobilis committed:
sd.cpp: use expm1() instead of (exp() - 1)

vmobilis committed:
sd.cpp: use 30k iterations in integrals

The result looks closer to long double.

vmobilis committed:
sd.cpp: use macros for pointers initialization

vmobilis committed:
sd.cpp: clean Heun sampler

Also, use switch() for buffers ceation.

vmobilis committed:
sd.cpp: clean Euler sampler

vmobilis committed:
sd.cpp: clean DPM++ 2S sampler

vmobilis committed:
sd.cpp: clean DPM++ 2S ancestral sampler

vmobilis committed:
sd.cpp: clean DPM++ 2M sampler

vmobilis committed:
sd.cpp: clean DPM++ 2M v.2 sampler

vmobilis committed:
sd.cpp: clean DPM2 sampler

vmobilis committed:
sd.cpp: clean LCM sampler

vmobilis committed:
sd.cpp: split original and modified euler_a

To view changes easier.

vmobilis committed:
sd.cpp: simplify buffers creation

Initialization with x_mat is needless.

vmobilis commented:
@vitoplantamura, hello, I think I cleaned the samplers (they all use same `sampler_history_buffer` now), but did not sync with new code base yet.
Merry Christmas! 🙂

vitoplantamura commented:
hi @vmobilis,

Very cool! It would be nice to have this PR merged into the project!

Let me know if you need help with the conflicts with the master branch.

Merry Christmas to you too!

Vito

vmobilis committed:
sd.cpp: moving samplers to separate file

vmobilis committed:
Create samplers.txt

vmobilis committed:
sd.cpp: sync with OnnxStream

vmobilis committed:
sd.cpp: use `steps` instead of `sigma.size() - 1`

vmobilis committed:
sd.cpp: move time measurement after sampler

2-step samplers also make denoising and need to be accounted.

vmobilis committed:
Merge branch 'master' into patch-2

vmobilis committed:
samplers.cpp: sync CFGDenoiser_CompVisDenoiser() call with master

vitoplantamura commented:
Happy New Year to you too, @vmobilis!

Yes, the changes are much more readable now. Maybe instead of having a cpp file included inline, we could consider an h file with a single inline function. That should be an easy fix. What do you think?

Vito

vmobilis commented:
@vitoplantamura, it might be good, I'll try to move all samplers-related code to a header.

vmobilis committed:
Rename samplers.cpp to samplers.h

vmobilis committed:
samplers.h: make samplers functions

vmobilis committed:
sd.cpp: move samplers to samplers.h

vmobilis committed:
Merge branch 'vitoplantamura:master' into patch-2

vmobilis committed:
samplers.h: rearrange process_sample() arguments

It seems to make smaller binary.

vmobilis committed:
sd.cpp: sync process_sample() arguments

Also move samplers.h inclusion lower and remove needless prototype.

vmobilis committed:
sd.cpp: remove needless steps == step variable

vmobilis commented:
@vitoplantamura, I think it's ready, I've tested all samplers and they work with SD, SDXL and SDXL Turbo.

The `ddpm`/`ddpm_a` samplers seems to work incorrectly, they produce very soft, blurry images.
I have no idea how to fix it, but on the other hand, with Turbo model results look interesting, so maybe it worth to not remove them.

Also, `dpm++2s` (non-ansestral) sampler seems to be broken too: I think it should produce result similar to Euler or DPM2, but it's completely different. Not bad, but different.
And I don't have Python-based Diffusers to say for sure whether it's wrong or not.

Samplers `dpm++3msde`, `ipndm_vo` and `lms` need x2 more steps than others.
Sampler `dpm++3msde_a` needs x10 steps or else it produces much noise.

Maybe it is better to mention it in help?

Also, images uploader on GitHub seems to be broken, I can not upload images and update examples in PR description.
I'll attach the results here and maybe update the examples later when I'll be able to upload them.
[images.zip](https://github.com/user-attachments/files/24683178/images.zip)

vitoplantamura commented:
Very nice and clean, I'll review the code ASAP!

Regarding images.zip, it would actually be helpful to have a table with all the images for reference. Perhaps we could create three large images (one for each model) with a thumbnail of each image and the name of the sampler (like those found by googling "stable diffusion samplers"). There must be an easy way to create this, maybe even a screenshot from an image viewer would do the trick... We could then link these three images in the main README, under the text "Options you're probably interested in", and, here, we could also add a text with the peculiarities of the samplers. What do you think?

Thank you,

Vito

vmobilis committed:
Merge branch 'vitoplantamura:master' into patch-2

vmobilis committed:
Upload samplers examples

vmobilis committed:
README.md: add samplers examples

vmobilis commented:
@vitoplantamura, I've added examples of samplers and script to merge them into a single image.

vitoplantamura commented:
Amazing contribution!

Thank you,

Vito
